### PR TITLE
[Misc] Update `gptq_marlin` to use new vLLMParameters

### PR DESCRIPTION
**Pull request by dsikka**
# Summary
- Updates the `gptq_marlin` parameters to use `vLLMParameters` to simplify linear layer weight loading
- Updates to add `PackedColumnParameter` to support packed parameters without row parallelism

### CODE REVIEW

File: vllm/model_executor/parameter.py [hunk 1]
```diff
@@ -160,18 +162,26 @@ def load_row_parallel_weight(self, loaded_weight: torch.Tensor):
...
+    row parallelism.
+    """
+    pass
+
+
+class GroupQuantScaleParameter(_ColumnvLLMParameter, RowvLLMParameter):
     """
     Parameter class for weight scales loaded for weights with
-    grouped quantization. Equivalent to ModelWeightParameter.
+    grouped quantization.  Uses both column and row parallelism.
```
Comment (mgoin) [inline vllm/model_executor/parameter.py:176]:
```suggestion
    grouped quantization. Uses both column and row parallelism.
```

File: vllm/model_executor/layers/quantization/gptq_marlin.py [hunk 2]
```diff
@@ -191,79 +197,66 @@ def create_weights(
...
+                output_dim=1,
+                packed_dim=1,
+                packed_factor=self.quant_config.pack_factor,
+                **qzeros_args)
+
+        else:
+            scales = GroupQuantScaleParameter(output_dim=1,
+                                              input_dim=0,
+                                              **weight_scale_args)
+            qzeros = PackedvLLMParameter(
```
[the remark this answers is not in the document]
Comment (dsikka) [reply vllm/model_executor/layers/quantization/gptq_marlin.py:254]: the Column and Row has to do with whether or not the parameter supports row and column parallelism, which changes the weight loading. They can all be 2D

File: vllm/model_executor/parameter.py [hunk 3]
```diff
@@ -232,14 +242,51 @@ def _load_into_shard_id(self, loaded_weight: torch.Tensor,
...
+    """
+    Parameter for model parameters which are packed on disk
+    and support column parllelism only. See PackedvLLMParameter
+    for more details on the packed properties.
+    """
+
+    def __init__(self,
+                 packed_factor: int,
+                 packed_dim: int,
+                 marlin_tile_size: Optional[int] = None,
```
[the remark this answers is not in the document]
Comment (dsikka) [reply vllm/model_executor/parameter.py:255]: wdym?

File: vllm/model_executor/layers/quantization/gptq_marlin.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgoin) [reply vllm/model_executor/layers/quantization/gptq_marlin.py:254]: Then can we remove the "vLLM"? It seems unrelated here if other parameters don't have vLLM in the name

File: vllm/model_executor/parameter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgoin) [reply vllm/model_executor/parameter.py:255]: Could we have a general `tile_size` for packed parameters (maybe named for a specific dim if that is assumed)? Nothing about the parameter mentions marlin except for this arg

[the remark this answers is not in the document]
Comment (dsikka) [reply vllm/model_executor/parameter.py:255]: The only use case where I've seen this is for marlin. Equivalent to the marlin_tile_size in the original weight loading flow: https://github.com/vllm-project/vllm/blob/5fb4a3f6785e3612bf1741f6e43a4184a37649c1/vllm/model_executor/layers/linear.py#L26

File: .buildkite/test-pipeline.yaml [hunk 4]
```diff
@@ -319,6 +319,17 @@ steps:
...
     - pytest -v -s -x lora/test_long_context.py
 
+- label: Weight Loading Multiple GPU Test
+  working_dir: "/vllm-workspace/tests"
+  num_gpus: 2
+  source_file_dependencies:
+  - vllm/
+  - tests/weight_loading
+  commands:
+    - export VLLM_WORKER_MULTIPROC_METHOD=spawn
```
Comment (robertgshaw2-neuralmagic) [inline .buildkite/test-pipeline.yaml:329]: you can drop this env variable

File: tests/weight_loading/models.txt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+gptq_marlin, robertgshaw2/zephyr-7b-beta-channelwise-gptq, main
+gptq_marlin, TheBloke/Llama-2-7B-GPTQ, main
```
Comment (robertgshaw2-neuralmagic) [inline tests/weight_loading/models.txt:2]: Trying to simplify this config, I think it would be better if it was just a list of model stubs

- you dont need the `gptq_marlin` in your test, vLLM will automatically detect
- we can just pick models which don't have to worry about revision

File: tests/weight_loading/models.txt [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+gptq_marlin, TheBloke/TinyLlama-1.1B-Chat-v1.0-GPTQ, main
+gptq_marlin, TheBloke/TinyLlama-1.1B-Chat-v1.0-GPTQ, gptq-8bit--1g-actorder_True
```
Comment (robertgshaw2-neuralmagic) [inline tests/weight_loading/models.txt:4]: could we add some of the compressed-tensors models too?

File: tests/weight_loading/test_weight_loading.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+def test_weight_loading(vllm_runner):
+    with vllm_runner(model_name=MODEL_NAME,
+                     revision=REVISION,
+                     dtype="half",
+                     quantization=QUANTIZATION,
+                     max_model_len=MAX_MODEL_LEN,
+                     tensor_parallel_size=2) as model:
+
+        output = model.generate_greedy("Hello world!", max_tokens=20)
```
Comment (robertgshaw2-neuralmagic) [inline tests/weight_loading/test_weight_loading.py:18]: Can you print the output so we have it in the logs?

File: tests/weight_loading/test_weight_loading.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+MODEL_NAME = os.environ.get("MODEL_NAME",
+                            "robertgshaw2/zephyr-7b-beta-channelwise-gptq")
+REVISION = os.environ.get("REVISION", "main")
+QUANTIZATION = os.environ.get("QUANTIZATION", "gptq_marlin")
+
+
+def test_weight_loading(vllm_runner):
+    with vllm_runner(model_name=MODEL_NAME,
+                     revision=REVISION,
+                     dtype="half",
```
Comment (robertgshaw2-neuralmagic) [inline tests/weight_loading/test_weight_loading.py:13]: use `auto` or don't specify so we get a diverse mix

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py [hunk 9]
```diff
@@ -105,7 +105,7 @@ def create_weights(self, layer: torch.nn.Module, input_size: int,
                 dtype=params_dtype,
             )
         }
-        if self.group_size == -1:
+        if not partition_scales:
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:108]: Can you explain this?

File: tests/weight_loading/models.txt (hunk 5 above)
Comment (dsikka) [reply tests/weight_loading/models.txt:2]: @robertgshaw2-neuralmagic 

2 things here:
1. I think we want test cases around running specific configs. This lets us do that. I understand it should be able to autodetect. But the goal is the detect the weight loading for a specific quantization config 
2. Sometimes the quantization used be autodetected based on compatibility and its not what we want to test (example: awq models running on awq-marlin)

File: tests/weight_loading/models.txt (hunk 6 above)
Comment (dsikka) [reply tests/weight_loading/models.txt:4]: yes

File: tests/weight_loading/test_weight_loading.py (hunk 7 above)
Comment (dsikka) [reply tests/weight_loading/test_weight_loading.py:18]: yes

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py (hunk 9 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:108]: better condition for when we're in both cases. I think it might end up being the same but more consistent with what ive seen in other quant config files

File: .buildkite/test-pipeline.yaml (hunk 4 above)
Comment (dsikka) [reply .buildkite/test-pipeline.yaml:329]: it was complaining about it before but can double check and remove

File: .buildkite/test-pipeline.yaml [hunk 10]
```diff
@@ -319,6 +319,16 @@ steps:
     - export VLLM_WORKER_MULTIPROC_METHOD=spawn
     - pytest -v -s -x lora/test_long_context.py
 
+- label: Weight Loading Multiple GPU Test
+  working_dir: "/vllm-workspace/tests"
+  num_gpus: 2
```
[the remark this answers is not in the document]
Comment (dsikka) [reply .buildkite/test-pipeline.yaml:324]: Yes. The point of the test is to make sure multi-gpu scenarios work well.
1 GPU scenarios are tested well already

File: tests/weight_loading/run_model_weight_loading_test.sh [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+do
+    LOCAL_SUCCESS=0
+    IFS=', ' read -r -a array <<< "$MODEL_CONFIG"
+    
+    echo "=== RUNNING MODEL: $MODEL_CONFIG ==="
+
+    export QUANTIZATION=${array[0]}
+    export MODEL_NAME=${array[1]}
+    export REVISION=${array[2]}
+    pytest -s weight_loading/test_weight_loading.py || LOCAL_SUCCESS=$?
```
[the remark this answers is not in the document]
Comment (dsikka) [reply tests/weight_loading/run_model_weight_loading_test.sh:16]: what functionality are you referring to?

[the remark this answers is not in the document]
Comment (mgoin) [reply tests/weight_loading/run_model_weight_loading_test.sh:16]: I was meaning why do we need a bash script. The config file could just as well be a list of tuples in the pytest file

**Review by mgoin — APPROVED** [review]

File: tests/weight_loading/run_model_weight_loading_test.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (dsikka) [reply tests/weight_loading/run_model_weight_loading_test.sh:16]: I think that would still be one process for all the models, which causes an issue with tp>1
bash script essentially lets us do one process per test/model to get around this problem
